### DISCUSSION TITLE
Comparison operator behaves differently for same inputs

### DISCUSSION THREAD

**Discussion by anakrish** [open-policy-agent/community#347] [OPA and Rego]
It seems that when a statement is a comparison expression then additional type-checks are performed causing the comparison to become invalid.

```
r  {
  [ [1, 2] == [2] ] # Equality is valid. Evaluates to false.
  [1, 2] == [2] # Equality is invalid. rego_type_error.
}
```

Is this behavior expected?

**Comment by anderseknert** [comment]
Hi @anakrish 👋 No, I think the first line should be a type error too. In general, the compiler will fail when it sees impossible equality comparisons performed, which is probably what you want it to do. As you've noticed, there are probably some cases that are not yet accounted for.

**Comment by anakrish** [comment]
Hi @anderseknert,

I wonder if it would be more appropriate for the second line to not error instead of the first line to error.
Both `[1, 2]` and `[2]` can be placed in a set `{ [1, 2], [2] }` and that would imply ability to compare the two values without raising errors.

Also, other comparison operators e.g `<` don't raise errors currently:
```
r  {
  [ [1, 2] <= [2] ] 
  [1, 2] <= [2]
}
# Above evaluates without errors, producing r = true
```

**Reply by srenatus** [reply]
Good insight. I believe the second == is rewritten to = in the compiler, and other type checking rules apply for that. The first one probably is not.

**Reply by anakrish** [reply]
Would you consider this a bug or do you think [Equality: Assignment, Comparison, and Unification](https://www.openpolicyagent.org/docs/latest/policy-language/#equality-assignment-comparison-and-unification) needs to be updated to reflect this behavior?
